src/s3/util/fileViewerManager.ts

```diff
@@ -168,20 +231,88 @@ export class S3FileViewerManager {
             }
         } else {
             //was activated from the explorer, need to get the file
+            const fileNode = uriOrNode
+            const headResponse = await fileNode.s3.getHeadObject({
+                bucketName: fileNode.bucket.name,
+                key: fileNode.file.key,
+            })
+            const type = mime.contentType(headResponse.ContentType!)
+            const charset = mime.charset(type as string)
+
             const fileLocation = await this.getFile(uriOrNode)
             if (!fileLocation) {
                 return
             }
+            const s3Uri = vscode.Uri.parse(fileLocation.fsPath)
+            let tab = this.activeTabs.get(fileLocation.fsPath)
+
+            if (!tab) {
+                tab = { fileUri: fileLocation, s3Uri, editor: undefined, s3FileNode: uriOrNode, type, charset } as S3Tab
+            }
+
+            if (charset != 'UTF-8') {
+                showOutputMessage(
+                    'Opening non-text file, please press enter on quickpick to continue.',
+                    this.outputChannel
+                )
+                tab.editor = await vscode.commands.executeCommand('workbench.action.quickOpen', tab.fileUri.fsPath)
+            } else {
+                tab.editor = await this.openFileGivenMode(tab, tab.fileUri, false)
+            }
 
-            const newTab = this.activeTabs.get(fileLocation.fsPath) ?? new S3Tab(fileLocation, uriOrNode)
-            await newTab.openFileInEditMode()
+            this.activeTabs.set(tab.fileUri.fsPath, tab)
+        }
+    }
+
+    /**
+     * Opens a given file on given tab and specified mode (read-only or edit mode)
+     *
+     * @param tab
+     * @param uri Uri to be opened will use the scheme attached to this
+     * @param preview boolean for argument to window.showTextDocument()
+     * @param workspace
+     * @returns
+     */
+    public async openFileGivenMode(
```

Comment:
Function name is misleading: you're not specifying a mode to open the file in. Since you're keying off the file scheme and since this function will always be called within the context of the fileViewerManager, you can probably just rename this to `openFile`.